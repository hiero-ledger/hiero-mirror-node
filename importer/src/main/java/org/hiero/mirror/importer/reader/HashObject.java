// SPDX-License-Identifier: Apache-2.0

package org.hiero.mirror.importer.reader;

import com.hedera.mirror.common.domain.DigestAlgorithm;
import java.io.IOException;
import lombok.EqualsAndHashCode;
import lombok.Value;
import org.hiero.mirror.importer.exception.InvalidStreamFileException;

@EqualsAndHashCode(callSuper = true)
@Value
public class HashObject extends AbstractStreamObject {

    int digestType;
    byte[] hash;

    public HashObject(ValidatedDataInputStream vdis, String sectionName, DigestAlgorithm digestAlgorithm) {
        super(vdis);

        try {
            digestType = vdis.readInt(digestAlgorithm.getType(), sectionName, "hash digest type");
            int hashLength = digestAlgorithm.getSize();
            hash = vdis.readLengthAndBytes(hashLength, hashLength, false, sectionName, "hash");
        } catch (IOException e) {
            throw new InvalidStreamFileException(e);
        }
    }

    public HashObject(ValidatedDataInputStream dis, DigestAlgorithm digestAlgorithm) {
        this(dis, null, digestAlgorithm);
    }

    protected HashObject(long classId, int classVersion, int digestType, byte[] hash) {
        super(classId, classVersion);
        this.digestType = digestType;
        this.hash = hash;
    }
}
