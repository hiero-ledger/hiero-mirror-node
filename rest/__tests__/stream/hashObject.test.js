// SPDX-License-Identifier: Apache-2.0

import HashObject from '../../stream/hashObject';

describe('HashObject', () => {
  let buffer;
  const classId = BigInt.asIntN(64, BigInt('0xf422da83a251741e'));
  const classVersion = 1;
  const digestType = 0x58ff811b;
  const hash = new Array(48).fill(0xef);

  beforeEach(() => {
    buffer = Buffer.from(
      [].concat(
        [0xf4, 0x22, 0xda, 0x83, 0xa2, 0x51, 0x74, 0x1e], // classId
        [0, 0, 0, 1], // classVersion
        [0x58, 0xff, 0x81, 0x1b], // digest type, sha-384
        [0, 0, 0, 0x30], // length, for sha-384, 48 bytes
        hash // 48-byte hash
      )
    );
  });

  test('getLength', () => {
    const expected = buffer.length;
    const hashObject = new HashObject(buffer);
    expect(hashObject.getLength()).toEqual(expected);
  });

  test('classId', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classId).toEqual(classId);
  });

  test('classVersion', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classVersion).toEqual(classVersion);
  });

  test('digestType', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.digestType).toEqual(digestType);
  });

  test('hash', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.hash).toEqual(Buffer.from(hash));
  });

  test('truncated buffer', () => {
    expect(() => new HashObject(buffer.slice(0, buffer.length - 4))).toThrowErrorMatchingSnapshot();
  });
});
