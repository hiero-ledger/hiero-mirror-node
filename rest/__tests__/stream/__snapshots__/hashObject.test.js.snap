// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HashObject truncated buffer 1`] = `"Error reading byte array, expect 48-byte data got 44-byte"`;
